### DISCUSSION TITLE
카카오링크에 대한 연구

### DISCUSSION THREAD

**Discussion by raravel** [storycraft/node-kakao#176] [General]
> 해당 토론에서는 얼마전부터 기존 코드로는 동작되지 않는 카카오링크에 대해 이야기합니다.
> 단순한 의견일지라도 한 줄 적어주시면 연구에 도움이 됩니다.

시험 버전은 `4.0.6 stable` 버전입니다.

정상 동작하는 카카오링크에 대한 정보는 [네이버 블로그 공유하기](https://blog.naver.com/ifthe1201/222212278758)를 기준으로 테스트했습니다.
(공유하기의 카카오링크 타입은FEED 입니다.)

> 버전 4에 들어와서 노드카카오는 Attachment 에 대한 정보를 JSON 그대로 표현하고 있습니다.
> 버전 3에 있던 [custom-attachment.ts](https://github.com/storycraft/node-kakao/blob/v3/src/talk/chat/attachment/custom-attachment.ts) 소스를 참고하여 각 속성에 대한 값을 유추할 것입니다.

카카오링크를 보내거나 받을때의 정보는 가장 상위에 `P(CustomInfo)`, `C(CustomContent)`, `K(KakaoLinkInfo)` 입니다.
여기서 `K`의 내부 값은 변경되어도 상관없습니다.

이외 `P` 또는 `C`의 내용이 변경되기라도 한다면 카카오링크는 전송되지 않습니다.
메시지 한 글자, 정보 하나라도 변경되면 올바르지 않은 카카오링크로 인식한다는 소리입니다. 오죽하면 썸네일 이미지 가로크기 정보인 `W`를 1 증가시켜도 보내지지 않습니다.

제가 생각하기엔 `P -> A` 객체의 내용이 검사의 열쇠가 된다고 생각됩니다.

버전 3의 [소스](https://github.com/storycraft/node-kakao/blob/v3/src/talk/chat/attachment/custom-attachment.ts#L865)를 보면 타입이 `boolean`으로 되어있었는데, 지금 보면 객체가 들어가더군요.

객체의 내용은 다음과 같습니다.

```json
{
	"version": 1,
	"code": "oophdFqvo3OsiUNuaKx8y920KMJ7w3N8Y/G4lmOxarg=",
	"createdAt": 1614825242755
}
```

`version`은 내버려둔다고 해도, `code`, `createAt`이 핵심같습니다.
`createAt`은 대충 카카오링크 생성시간같고요. `Date.now()`의 값입니다.

`code`는 보시다시피 base64 인코딩이 되어있는데 디코딩해보면 항상 다른 32바이트 길이의 값이 나옵니다.
처음에는 이 값을 분석하면 될 것 같았는데, 지금보니 이 값은 그냥 해시값인 것 같습니다.

해시값을 가지고 서버에서 유효성을 검사하는 것 같고, 그러면 원하는 정보를 토대로 유효성 OK를 받는 카카오링크는 어떻게 동작해야될지 알아봐야 합니다.

저도 틈날 때마다 연구하고 진척이 있으면 업데이트 할테니 여러분도 카카오링크를 보내는 방법을 업데이트 해주시면 감사하겠습니다.

# 틈날 때마다 연구하고 피드를 업데이트 중이니 가장 마지막에 업데이트 된순으로 보시려면 오른쪽 아래 Newest로 정렬해서 봐주세요.

**Comment by ghost** [comment]
https://developers.kakao.com 여기서 카카오링크 템플릿을 생성할 수 있는데, 여기서 힌트가 나올 수 도 있지 않을까 싶습니다.

**Reply by ghost** [reply]
http://kapi.kakao.com/v2/api/kakaolink/talk/template/validate 카카오링크가 올바른지 확인하는 링크 같습니다.

**Reply by raravel** [reply]
감사합니다. 그러면 카카오링크 템플릿을 사용하겠다는 요청을 보내야 유효성검사를 지나칠 수 있겠네요.

**Reply by ghost** [reply]
근데 신기하네요. 카카오톡 어플에도 Ad는 Boolean으로 되어있는데, 갑자기 한순간에 객체가 되다니..

**Reply by ghost** [reply]
> http://kapi.kakao.com/v2/api/kakaolink/talk/template/validate 카카오링크가 올바른지 확인하는 링크 같습니다.

확인해보니 카카오톡 공유에서 카카오링크 템플릿을 공유할때 사용되는 링크입니다.

**Comment by raravel** [comment]
`P -> A -> code` 의 해시 알고리즘은 SHA256 입니다.
해시 값에는 `createAt` 값이 포함됩니다.

**Reply by ghost** [reply]
좋은 소식이네요.

**Comment by raravel** [comment]
일단은 단순히 노카로 보내는 방법을 빠른 시일내에 성공할 수 있을 것 같지 않아, 브라우저에서 보내는대로의 동작을 따라하는 모듈을 만들고 있습니다.

분석 중, 채팅방 목록을 가져오는데, `id` 값에 어떠한 알 수 없는 값이 들어갔습니다.
이 ID값이 어떤 값인지 알면, 채팅방 제목인 `title` 값으로 가져오는 것 보다 더 정확하게 보낼 수 있을 것 같은데 노카에서는 어떻게 채널 정보로 이 값을 만들 수 있을까요.

`8GwlB64L9AVey1JYd5q0jDXvpe203qzJlwAiy00vNEMPxgnRbKGmk738zo2OrQav` 이런 값인데 base64 인코딩 된 거 같아서 디코딩해보니 48바이트의 해시값이 나왔습니다. 형태로 보아 SHA-384일 확률이 높습니다.

혹시 알고 계신분은 답변주시면 감사하겠습니다.

**Reply by raravel** [reply]
![image](https://user-images.githubusercontent.com/28672888/110901744-22395a00-8348-11eb-97d1-67cd598698db.png)
8일 전 이 글 등록당시의 데이터가지고 현재도 보내지는 걸 확인했습니다.

**Comment by raravel** [comment]
카카오링크 모듈을 따로 만들었습니다.
노카로 보낼 수 있는 방법을 찾기 전까진 이 모듈을 사용하려고 합니다.
물론, 이 모듈을 사용해서 보낸 패킷을 저장하고 있다가 보낸다면 노카에서도 카링을 부분적으로 보낼 수 있다고 볼 수 있습니다.

모듈은 위에서 말한 것과 같이 브라우저에서 동작하는 내용을 자동으로 할 수 있도록 포팅한 것이며, 그만큼 효율이 노카로 보내는 것보다 떨어집니다.
모듈의 소스 및 사용법은 [이곳](https://github.com/raravel/kaling.js)에서 확인하실 수 있습니다.

**Comment by raravel** [comment]
카카오톡 앱 `9.2.6` 버전을 디컴파일 하여 분석해봤습니다.
안드로이드 개발 및 자바/코틀린에 대해서 Hello World도 못 찍는 저라서 전부 가정형으로 설명드리는 것 양해바랍니다.
** 또한 처음부터 전부 틀린 내용일 수 있습니다. **

[REDACTED] 패키지에 [REDACTED] 클래스가 존재하고, 이것이 지금까지 저희가 써왔던 카카오링크 전송 정보를 가지고 있는 클래스입니다. 이전에 말했다 시피, 기존에 `Boolean` 형태였던 `A` 키는 JSON 형태가 되고 기존 A는 AD가 되었습니다.


분석한 바로는 [REDACTED] 패키지에 카카오링크 관련 검사 메서드인 [REDACTED] 함수가 유효성검사하는 함수인 것 같습니다.
아래로는 유효성 검사하는 루틴들을 차례대로 설명한 것입니다.

### 1
먼저 `P > VA`의 클라이언트 버전을 검사합니다. 앱 버전 정보(9.2.6) 과 VA(보통 6.0.0) 을 비교해서 더 작으면 반환 결과 플래그값을 TRUE로 세팅하지 않습니다.
그래서 어? 앱 버전을 맞춰줘야 하는 건가? 라고 할 수 있지만, 그 이전에 빈값이 아니면 TRUE로 반환합니다.
??? 일단 값만 있으면 된다고 하니 다행입니다.

### 2
[REDACTED]\(P) 항목의 내용중 subType(TP), message(ME), serviceId(SID), docId(DID)들을 코틀린 isBlank함수 사용, 빈값이 아닌지 검사합니다.
이로써 저희는 TP, ME, SID, DID는 필수항복이란 것을 확인했습니다.

Content(C) 값이 존재할 땐, 각각의 유효성검사를 따로 한다는 것도 알았습니다. FeedContent든 ListContent든 Content 추상 클래스를 상속받습니다. 그 중 유효성 검사를 해 `boolean`을 반환하는 추상메서드가 존재합니다.


### 3
verified(KV)의 값이 FALSE라면, forwardable(FW) 또한 FALSE여야 합니다.
근데, KV의 값이 TRUE라면 FALSE를 반환합니다. (?????)


### 4
subType(TP) 의 값이 `carousel` 이어야 합니다. (??!?!)


쓰면서 여기까지 알아보니 이 검사루틴은 공유관리자에서 사용하는 것이고, 특정 채팅 타입을 검사하기 위한 루틴인 것 같습니다.
lock(LOCK)값이 TRUE여야 하고, verified(KV)가 FALSE면서 bt(BT)는 TRUE여야 한답니다.

결과적으로, 허탕친 것을 알았습니다. 그러나 1번과 2번 루틴은 그럴싸한 검사방식이니 아주 허탕은 아니라고 심심한 본인위로를 합니다.
다시 분석하고 피드를 업데이트 하겠습니다.

-----
Note: 해당 글은 @storycraft 에 의해 일부 내용이 수정되었습니다.

**Comment by raravel** [comment]
저희에게 중요한 것은 `P > A > code`입니다.

샵검색 기능이 `FeedAttachment`를 사용하는 것으로 바뀐지 얼마 안 된 것을 아실겁니다.
이를 이용해 후킹해본 결과 로코 전송시 이미 `code`값은 부여된 것을 알 수 있었습니다.

이는 서버에서 유효성 검사를 등록하는 게 아니라 유효성 코드를 부여한 상황에서 전송한다는 것을 알 수 있습니다.

다음은 후킹한 로그 중 로코 전송시에 데이터이니 관심있으신 분은 보시기 바랍니다.

```
[TransLayer:LocoHeader { packetId = 689209641, status = 0, method ='MSG', bodyLength = 1648 }] [Body:{chatLog={logId=2495342749620174848, chatId=279736756166680, type=71, authorId=333120445, message=샵검색 : #asdgasdg, sendAt=1618405236, attachment={"P":{"TP":"List","ME":"샵검색 : #asdgasdg","SID":"sharp","DID":"https://m.search.daum.net/kakao?w=tot&q=asdgasdg&DA=SH1&rtmaxcoll=IVR","VA":"8.6.5","VI":"8.6.5","VM":"2.6.9","VW":"3.0.8","RF":"sharp_search","BT":false,"FW":true,"A":{"version":1,"code":"NCv72Pf66jpnasSuqEwr1ezxllgDNRf5CW5xpAJ7CFY=","createdAt":1618405236370}},"C":{"HD":{"TD":{"T":"asdgasdg"},"L":{"LPC":"https://m.search.daum.net/kakao?w=tot&q=asdgasdg&DA=SH1&rtmaxcoll=IVR","LMO":"https://m.search.daum.net/kakao?w=tot&q=asdgasdg&DA=SH1&rtmaxcoll=IVR","LAA":false},"ICO":"https://search1.daumcdn.net/search/statics/img/ki/ico_sharpsearch.png","TOP":true},"BUL":[{"BU":{"T":"검색 결과 더보기","SR":"both","TP":"more","HL":false},"L":{"LPC":"https://m.search.daum.net/kakao?w=tot&q=asdgasdg&DA=SH1&rtmaxcoll=IVR","LMO":"https://m.search.daum.net/kakao?w=tot&q=asdgasdg&DA=SH1&rtmaxcoll=IVR","LAA":false,"LAD":"검색 결과 더보기"}}],"ITL":[{"L":{"LPC":"http://blog.daum.net/sor3244","LMO":"http://blog.daum.net/sor3244","LAA":false},"TD":{"T":"asdgasdg","D":"http://blog.daum.net/sor3244"},"KMA":false},{"L":{"LPC":"http://blog.daum.net/fcgzfgasfdga","LMO":"http://blog.daum.net/fcgzfgasfdga","LAA":false},"TD":{"T":"asdgasdg","D":"http://blog.daum.net/fcgzfgasfdga"},"KMA":false},{"L":{"LPC":"http://blog.daum.net/ddasdfasdg","LMO":"http://blog.daum.net/ddasdfasdg","LAA":false},"TD":{"T":"asdgasdg","D":"http://blog.daum.net/ddasdfasdg"},"KMA":false}]}}, msgId=1350125389, prevId=2495342651548983296}, chatId=279736756166680, pushAlert=false, noSeen=false}]
```

![image](https://user-images.githubusercontent.com/28672888/114715475-10473e80-9d6e-11eb-9e5e-95a4b70f8f9d.png)
